stan/math/prim/mat/matrix_addons.h

```diff
@@ -0,0 +1,206 @@
+/**
+ * Structure to determine whether input struct has member named "d_".
+ * This is used to differentiate between var and fvar<T> without
+ * depending on external code
+ */
+template <typename T> 
+struct Hasd_ { 
+  typedef typename std::remove_pointer<T>::type decay_type;
+    struct Fallback { int d_; };
+    struct DerivedType : decay_type, Fallback { };
+
+    template<typename C, C> struct ChT; 
+    template<typename C> static char (&f(ChT<int Fallback::*, &C::d_>*))[1]; 
+    template<typename C> static char (&f(...))[2]; 
+
+    static bool const value = sizeof(f<DerivedType>(0)) == 2;
+}; 
```

Comment:
Neat! Reviewing this is certainly above my paygrade, but would you mind doc'ing more about what's going on here? For instance the line

```cpp
template<typename C> static char (&f(ChT<int Fallback::*, &C::d_>*))[1]
```

That's kinda wild looking, not sure how to read that

Reply:
> That's kinda wild looking, not sure how to read that

I agree.  We should assume the reader is good at C++, but this one could use some unpacking.  At the very least, template parameters should get documentation.

In general, we should favor includes to directly including declarations like this as they're easier to maintain.  In some cases, it can help to just include declarations directly, but I'm not sure this is one of them.  

Some typedefs might help.

I can break it down piecewise to be declaring the function `f` to apply to some complex pointer to `ChT` and return something that can be dereferenced and applied as an array to index 1 to produce a `char`.  

The bit I couldn't parse was `int Fallback::*`.  What is that?

Reply:
Googling, Fallback::* is a pointer the member function which is declared as an int so doing implicit conversion.

idt an int is always guaranteed to be big enough to hold any pointer (from [here](https://stackoverflow.com/a/25225033/2269255)). So maybe that should be something like

```cpp
template<typename C> static char 
(&f(ChT<uintptr_t reinterpret_cast<uintptr_t>(Fallback::*),
 &C::d_>*))[1]; 
```
Where `d_` is also a `uintptr_t` (?)

Reply:
I can't take the credit for this clever piece of work. I took the approach from [this blog post](https://cpptalk.wordpress.com/2009/09/12/substitution-failure-is-not-an-error-2/).

I had to resort to something like this because you can't ```#include``` external headers in the ```matrix_addons``` file. The other alternative would be to include the appropriate header before declaring the add-on in ```prim/mat/Eigen```. But I don't think that would be a good approach since it would create a more complex requirement on the order of header includes

Reply:
Huh! Pretty cool blog, may make me a bit too clever if I read his stuff

> I had to resort to something like this because you can't #include external headers in the matrix_addons file. The other alternative would be to include the appropriate header before declaring the add-on in prim/mat/Eigen. But I don't think that would be a good approach since it would create a more complex requirement on the order of header includes

Even for standard library headers? Cuz it looks kind of like this is doing something like `enable_if_t<is_fvar<T>, blah blah>`

Reply:
^By "Even for standard library headers" I meant, is including a standard library header before declaring the add on fall into the realm of "too complex". If we doc the above well then I'm fine with it, but a little bit of complexity in the header order is also a fair trade for reduced code complexity here

Reply:
If we wanted to catch people who do the order in the wrong way we could also have do something like define `STAN_MATH_EIGEN_ADDONS_CHECK` in Eigen.hpp, then check and throw something nice during complication if the header wasn't included in the right order

Reply:
If we use other math headers to differentiate between ```fvar<T>``` and ```var``` then we'd have to include headers from both ```rev``` and ```fwd``` when ```prim/mat/fun/Eigen.hpp``` is included, which could make things a bit messy.

I've updated the doc on that class, but let me know if there's a better approach I could be using here

Reply:
Turns out this method has been superseded by a much [simpler approach using ```void_t```](https://en.cppreference.com/w/cpp/types/void_t). Makes the code a bit more readable!

Reply:
@andrjohns just thinking out loud... if you're having to redefine traits that we already have, would it just be easier to not have this as a plugin and just create a view? I think you're running into the dependency of: the Eigen plugin needs to know about Math's autodiff types AND Math's traits are set up so that we have specialization for Eigen's types.

I think that problem could be avoided by writing a view where we already know about both Eigen and Math.

Reply:
@syclik so my interest in this PR is a bit selfish. I'm working on a PR right now that would make the OpenCL API look v similar to what Andrew has here.

It's a good bit down the road, but in the stan3 compiler Sean wants to be able to generate code that directly work with `matrix_cl`s (instead of passing back and forth inside of our normal functions). If we can line up the regular api and OpenCL api's I'm pretty sure we could do something like template the signatures to accept either Eigen matrices or matrix_cl types and then it would "just work"

Reply:
> if you're having to redefine traits that we already have, would it just be easier to not have this as a plugin and just create a view?

I would agree if the added code was substantial, but it's ended up only being a few lines:
```
template<class, class = void>
struct Hasd_ : std::false_type
{ };
template<class T>
struct Hasd_<T, decltype((void)(T::d_))> : std::true_type
{ };
```
So I think we should leave it with the plugin, to be honest.

Reply:
Cool. (I couldn't tell how much duplicated effort there was.)